src/tests/testClpMap.cc

```diff
@@ -0,0 +1,128 @@
+{
+    for (auto j = startWith; j < startWith + count; ++j)
+    {
+        CPPUNIT_ASSERT_EQUAL(true, m.add(std::to_string(j), j, ttl));
+    }
+}
+
+
+void
+testClpMap::setUp()
+{
+    squid_curtime = time(nullptr);
+}
+
+void
+testClpMap::testPutGetDelete()
+{
+    TestMap m(1024);
+    addData(m, 10, 0, 10);
+    CPPUNIT_ASSERT_EQUAL(static_cast<const int *>(nullptr), m.get("notthere"));
+    CPPUNIT_ASSERT_EQUAL(1, *(m.get("1")));
+    CPPUNIT_ASSERT_EQUAL(9, *(m.get("9")));
+    m.add("1", 99);
+    CPPUNIT_ASSERT_EQUAL(99, *(m.get("1")));
+    m.del("1");
+    CPPUNIT_ASSERT_EQUAL(static_cast<const int *>(nullptr), m.get("1"));
+}
+
+void testClpMap::testEntries()
+{
+    {
+        TestMap m(10*1024*1024, 10);
+        addData(m, 10, 10, 10);
+        CPPUNIT_ASSERT_EQUAL(static_cast<size_t>(10), m.entries());
+        m.add("foo", 0);
+        CPPUNIT_ASSERT_EQUAL(static_cast<size_t>(11), m.entries());
+    }
+    {
+        TestMap m(1024, 5);
+        addData(m, 1000, 0, 10);
+        CPPUNIT_ASSERT(m.entries() < 1000);
+    }
+}
+
+void
+testClpMap::testMemoryCounter()
+{
+    CPPUNIT_ASSERT_EQUAL(sizeof(int), static_cast<size_t>(DefaultMemoryUsage(int())));
+    CPPUNIT_ASSERT_EQUAL(sizeof(int32_t), static_cast<size_t>(DefaultMemoryUsage(int32_t())));
+    CPPUNIT_ASSERT_EQUAL(sizeof(int64_t), static_cast<size_t>(DefaultMemoryUsage(int64_t())));
+    CPPUNIT_ASSERT_EQUAL(sizeof(char), static_cast<size_t>(DefaultMemoryUsage(char())));
+    char str[10];
+    CPPUNIT_ASSERT_EQUAL(sizeof(str), static_cast<size_t>(DefaultMemoryUsage(str)));
+    CPPUNIT_ASSERT_EQUAL(sizeof(std::string), static_cast<size_t>(DefaultMemoryUsage(std::string())));
+}
+
+void
+testClpMap::testConstructor()
+{
+    TestMap nilA(0);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilA.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilA.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilA.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), nilA.entries());
+
+    TestMap nilB(0, 0);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilB.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilB.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilB.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), nilB.entries());
+
+    TestMap emptyC(1);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1), emptyC.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1), emptyC.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), emptyC.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), emptyC.entries());
+
+    TestMap emptyD(1024);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1024), emptyD.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1024), emptyD.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), emptyD.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), emptyD.entries());
+}
+
+void
+testClpMap::testSetMemLimit()
+{
+    TestMap m(2048);
+    addData(m, 1000, 0, 10);
+    auto testEntriesBefore = m.entries();
+    m.setMemLimit(1024);
+    CPPUNIT_ASSERT(testEntriesBefore > m.entries());
+}
```

Comment:
It would be nice to tell that the map can grow after purging entries. Squid had bugs in similar conditions:

```suggestion

    const auto entriesAfterPurge = m.entries(); // TODO: Move and use in the above assertion
    m.setMemLimit(m.memLimit() * 2);
    // overflow the map with entries again to make sure it can grow after purging
    addData(m, m.memLimit() / sizeof(int), 0, 10); // TODO: Extract into overflowMap()
    CPPUNIT_ASSERT(entriesAfterPurge < m.entries());
}
```

It would be nice to make these checks more strict/accurate, but that requires more code.

P.S. The above TODOs are in scope. I just could not fit those changes into one change request.